lib/src/source/hosted.dart

```diff
@@ -537,6 +547,183 @@ class HostedSource extends CachedSource {
+          return response.body;
+        });
+      });
+      final decoded = jsonDecode(bodyText);
+      if (decoded is! Map<String, dynamic>) {
+        throw FormatException('security advisories must be a mapping');
+      }
+      body = decoded;
+      result = _parseAdvisories(decoded, ref.name);
+    } on FormatException catch (error, stackTrace) {
+      _throwFriendlyError(error, stackTrace, packageName, hostedUrl);
+    }
+
+    // Cache the response on disk.
+    // Don't cache overly big responses.
+    if (bodyText.length < 5000 * 1024) {
+      await _cacheAdvisoriesResponse(body, ref, cache);
+    }
+    return result;
+  }
+
+  List<Advisory>? _parseAdvisories(Map body, String packageName) {
+    final advisoryIds = <String>[];
+    final advisoriesResult = <Advisory>[];
+    final affectedVersions = <String, List<String>>{};
+
+    final advisories = body['advisories'];
+    if (advisories is! List) {
+      throw FormatException('advisories must be a list');
+    }
+
+    final advisoriesUpdated = body['advisoriesUpdated'];
+    if (advisoriesUpdated is! String) {
+      throw FormatException('advisoriesUpdated must be a String');
+    }
+
+    for (final advisory in advisories) {
+      if (advisory is! Map) {
+        throw FormatException('advisory must be a map');
+      }
+
+      final id = advisory['id'];
+      if (id is! String) {
+        throw FormatException('id must be a String');
+      }
+      advisoryIds.add(id);
+
+      final affectedPackages = advisory['affected'];
+      if (affectedPackages is! List) {
+        throw FormatException('affectedPackages must be a list');
+      }
+
+      final affectedPkg = affectedPackages.firstWhereOrNull((element) {
+        if (element is! Map) {
+          throw FormatException('affectedPackage must be a map');
+        }
+        final pkg = element['package'];
+        if (pkg is! Map) {
+          throw FormatException('package must be a map');
+        }
+        final name = pkg['name'];
+        if (name is! String) {
+          throw FormatException('package name must be a String');
+        }
+        if (name == packageName) {
+          final ecosystem = pkg['ecosystem'];
+          if (ecosystem is! String) {
+            throw FormatException('ecosystem must be a String');
+          }
+          return ecosystem.toLowerCase() == 'pub';
+        }
+        return false;
+      }) as Map?;
+
+      if (affectedPkg == null) {
+        throw FormatException(
+          'Advisory $id does not contain $packageName among its affected packages.',
+        );
+      }
+
+      final versions = affectedPkg['versions'];
+      if (versions is! List) {
+        throw FormatException('package versions must be a list');
+      }
+
+      if (versions.any((element) => element is! String)) {
+        throw FormatException('package version elements must be a string');
+      }
+
+      advisoriesResult
+          .add(Advisory(id, [...versions.map((e) => e.toString())]));
```

Comment:
Why do we call `toString` when we know from above it is a String?

Should we parse these versions here and store them as version objects? (With a formatException when a version is not a version)